Modular Gap Transfer Models

dschwen opened the pull request:
Closes #19229 
Refs #19223

dschwen committed:
Add coverage test (#19223)

dschwen committed:
First draft (#19229)

dschwen committed:
Add radiative model (#19229)

dschwen committed:
Refactor interface UO (#19229)

dschwen committed:
Add material dependencies (#19229)

dschwen committed:
AD and fix typo (#19229)

dschwen committed:
Add dummy docs (#19229)

dschwen committed:
Add radiative transfer test, document code (#19229)

dschwen commented:
@lindsayad the test fails with non-sparse AD. Any idea why? I'm doing some funky stuff with passing dependencies on variables and material properties from coupled user objects through to a constraint. I'm wondering if this could be related to neighbor values/properties not being properly initialized. But then again, why does it converge to a wrong solution rather than not converge at all?

dschwen commented:
Actually this makes no sense. The material properties that the failing test coupled in are constant number literals provided in the input. Their derivatives should be zero anyways!

dschwen committed:
Skip test, add dependency (#19229)

dschwen commented on framework/include/materials/TwoMaterialPropertyInterface.h:
```diff
@@ -114,7 +114,10 @@ TwoMaterialPropertyInterface::getNeighborADMaterialProperty(const std::string &
     return *default_property;
   else
+  {
+    _material_property_dependencies.insert(_material_data->getPropertyId(prop_name));
```
Any idea why no dependency was added for neighbor AD properties?!

lindsayad replied:
Because someone was a bad boy! (It was me. Added it for literally everything else except this. Don't know what I was doing!)

lindsayad commented:
> @lindsayad the test fails with non-sparse AD.

what is the test?

lindsayad commented:
With the dependency addition, does the test still exodiff?

dschwen commented:
I'll try that. Right now I'm skipping the test.

On Mon, Nov 1, 2021, 7:08 PM Alex Lindsay ***@***.***> wrote:

> With the dependency addition, does the test still exodiff?
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/19236#issuecomment-957011795>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABRMPV72CETAV63Z7XWLD3UJ422JANCNFSM5G5ES34Q>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

lindsayad commented:
Yea just would be good to get the bug ironed out and covered in testing

lindsayad commented:
Doesn't have to be in this PR since you didn't make the bug ... but soon

dschwen commented:
@lindsayad the material property dependence fix does not fix the diff for non-sparse AD :-/

dschwen commented:
Here's a link to the failed test https://civet.inl.gov/job/870612/

lindsayad commented:
I thought you added that dependency ~4 hours ago, but that job finished ~7 hours ago?

dschwen commented:
The linked test is not the latest Civet test but it shows the non-sparse failure nonetheless. The latest civet test has the dependency, but also skips non-sparse testing. The combination of non-sparse and dependency fix was only tested locally on my machine and showed the exact same result as the Civet test I linked to.

dschwen committed:
Address comments, add displaced test (#19229)

lindsayad commented:
@dschwen I think you need some more branches

lindsayad committed:
New class works fine with nonsparse local AD

With global AD indexing we always force a full preconditioning matrix
because we have no way to know what global indices correspond to what
variables without building a map. However, with local numbering we do
know that information so we don't force a full preconditioning matrix.
Consequently convergence will be sensitive to what a user specifies in
their input. In this case, no preconditioning matrix is specified so for
the nonsparse case you get a block diagonal. With different matrices,
the global and local setups converge differently and it just so happened
that the global converged to a lower norm than the local case. By
scaling the temperature and LM we force both variables to finish the
simulation converged, and this gives us consistent results across AD
types
